<a href="https://colab.research.google.com/github/Jmd1704/2025-segundo-semestre-ia-economia/blob/main/Practica_limpieza_Juan_Medina_Estudiante.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# **Parte práctica parcial: Preparación y limpieza de datos**🧹

**IMPORTANTE**: Guarda una copia de este notebook en tu Google Drive o computador.

**Nombre estudiante:**

**Forma de entrega**

Jupyter Notebook publicado en su cuenta de Github con el nombre “Practica_limpieza_nombre_apellido_estudiante.ipynb”. Máximo a las 9 a.m., debes enviar link del notebook al correo lina.castro6@uexternado.edu.co, de lo contrario, no será tenido en cuenta.

**Instrucciones Generales**

- Puedes utilizar tus apuntes y notebooks y entrar al repositorio del curso. Sin embargo, **NO puedes utilizar IA para desarrollar tus respuestas**. **En caso de descubrir que estás usando IA** para el desarrollo de los ejercicios, **tu nota** de este segundo parcial **será cero**.
- Completa el código en las celdas marcadas con `### TU CÓDIGO AQUÍ ###`. Puedes añadir más celdas si lo requieres.
- Después de cada punto, ejecuta la celda de `assert` que se proporciona. Si no aparece ningún error, ¡tu respuesta es correcta!
- No modifiques las celdas de `assert` ni las celdas de Markdown.

**Objetivos de Aprendizaje**

1. **Aplicar técnicas de importación y consolidación de múltiples fuentes de datos** utilizando la librería `pandas`.
2.  **Ejecutar un proceso completo de limpieza de datos**, incluyendo el manejo de valores nulos, duplicados y la estandarización de tipos de datos.
3.  **Preparar un conjunto de datos real (GEIH del DANE)** para un análisis económico, sentando las bases para futuros modelos de machine learning.

# Ejercicio 1: Importar y unir datos (20%)

El DANE publica los microdatos de la GEIH mensualmente. Tu primera tarea es cargar los datos de los primeros seis meses de 2025 y armar una sola base.

**Los archivos necesarios los encuentras en el repositorio del curso, carpeta "datasets", carpeta "geih_2025".**

1. Usa un ciclo "for" para abrir los archivos CSV, ten en cuenta que estas bases usan como separador ";" y encoding='latin1'. La variable 'DPTO' y 'RAMA4D_R4' debes cargarlas como string.

In [ ]:
### TU CÓDIGO AQUÍ ###
import pandas as pd
import os

directory = '/content'
csv_files = [f for f in os.listdir(directory) if f.endswith('.csv')]
dfs = []

for file in csv_files:
    file_path = os.path.join(directory, file)
    df = pd.read_csv(file_path, sep=";", encoding="latin1",dtype={'DPTO': str, 'RAMA4D_R4': str})
    dfs.append(df)

/tmp/ipython-input-69537269.py:11: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, sep=";", encoding="latin1",dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-69537269.py:11: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, sep=";", encoding="latin1",dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-69537269.py:11: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, sep=";", encoding="latin1",dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-69537269.py:11: DtypeWarning: Columns (71,132,176,185,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, sep=";", encoding="latin1",dtype={'DPTO': str, 'RAMA4D

2. Une todos los dataframes para obtener una base de la GEIH con toda la información de enero a junio de 2025. El nuevo dataframe debe llamarse `df_geih`

In [ ]:
### TU CÓDIGO AQUÍ ###
df_geih = pd.concat(dfs, ignore_index=True)
print("DataFrame final:", df_geih.shape)
df_geih.head()


DataFrame final: (408423, 245)


,PERIODO,MES,PER,DIRECTORIO,SECUENCIA_P,ORDEN,HOGAR,REGIS,AREA,CLASE,...,P1881,P1881S1,P1882,P7240,P7240S1,OCI,INGLABO,RAMA2D_R4,RAMA4D_R4,OFICIO_C8
0,20250522,5,2025,8207971,1,1,1,10,NaN,2,...,14.0,NaN,NaN,3.0,NaN,1.0,3400000.0,96.0,9609,5169.0
1,20250522,5,2025,8207971,1,2,1,10,NaN,2,...,14.0,NaN,NaN,3.0,NaN,1.0,1400000.0,96.0,9609,5169.0
2,20250522,5,2025,8207972,1,1,1,10,NaN,2,...,14.0,NaN,NaN,5.0,NaN,1.0,750000.0,49.0,4921,8321.0
3,20250522,5,2025,8207972,1,2,1,10,NaN,2,...,2.0,NaN,40.0,3.0,NaN,1.0,840000.0,47.0,4773,5223.0
4,20250522,5,2025,8207973,1,1,1,10,NaN,2,...,3.0,NaN,10.0,3.0,NaN,1.0,600000.0,41.0,4111,9313.0


3. Reinicia el índice de `df_geih`. Asegúrate de que el índice antiguo no se añada como una nueva columna.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_geih = df_geih.reset_index(drop=True)
df_geih.head()


,PERIODO,MES,PER,DIRECTORIO,SECUENCIA_P,ORDEN,HOGAR,REGIS,AREA,CLASE,...,P1881,P1881S1,P1882,P7240,P7240S1,OCI,INGLABO,RAMA2D_R4,RAMA4D_R4,OFICIO_C8
0,20250522,5,2025,8207971,1,1,1,10,NaN,2,...,14.0,NaN,NaN,3.0,NaN,1.0,3400000.0,96.0,9609,5169.0
1,20250522,5,2025,8207971,1,2,1,10,NaN,2,...,14.0,NaN,NaN,3.0,NaN,1.0,1400000.0,96.0,9609,5169.0
2,20250522,5,2025,8207972,1,1,1,10,NaN,2,...,14.0,NaN,NaN,5.0,NaN,1.0,750000.0,49.0,4921,8321.0
3,20250522,5,2025,8207972,1,2,1,10,NaN,2,...,2.0,NaN,40.0,3.0,NaN,1.0,840000.0,47.0,4773,5223.0
4,20250522,5,2025,8207973,1,1,1,10,NaN,2,...,3.0,NaN,10.0,3.0,NaN,1.0,600000.0,41.0,4111,9313.0


In [ ]:
# Celda de verificación (no modificar)
assert df_geih.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_geih.shape[0]}."
assert df_geih.shape[1] == 245, f"Error en las columnas. Se esperaban 245, pero se encontraron {df_geih.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
assert 'index' not in df_geih.columns, "El índice antiguo no debe ser una columna."
assert df_geih.index.is_unique, "El índice debe ser único después de reiniciarlo."
print(f"¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene {df_geih.shape[0]} filas y {df_geih.shape[1]} columnas.")

¡Las dimensiones del DataFrame son correctas! ✅
¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene 408423 filas y 245 columnas.


# Ejercicio 2:  Seleccionar y renombrar columnas (5%)

La GEIH contiene cientos de variables. Para nuestro análisis del mercado laboral, nos enfocaremos en un subconjunto.

Adicionalmente, los nombres de las columnas en la GEIH son códigos. Para hacer el análisis más intuitivo, vamos a renombrarlas.

1. Crea un nuevo DataFrame llamado `df_empleo` que contenga únicamente las columnas que están en la lista `columnas_interes`.

In [ ]:
columnas_interes = ['DIRECTORIO', 'SECUENCIA_P', 'ORDEN', 'MES', 'CLASE', 'DPTO', 'FEX_C18', 'PT', 'P3271', 'P6040', 'P6080', 'P3043', 'OCI', 'P6430', 'RAMA4D_R4', 'INGLABO']

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_geih[columnas_interes]

In [ ]:
# Celda de verificación (no modificar)
assert len(df_empleo.columns) == 16, "El número de columnas es incorrecto."
assert all(col in df_empleo.columns for col in columnas_interes), "Faltan columnas o los nombres son incorrectos."
print("¡Selección de columnas completada con éxito! ✅")
print("Dimensiones del nuevo DataFrame:", df_empleo.shape)

¡Selección de columnas completada con éxito! ✅
Dimensiones del nuevo DataFrame: (408423, 16)


2. Renombra las columnas del DataFrame `df_empleo` utilizando el diccionario `mapa_nombres`.

In [ ]:
mapa_nombres = {
'DIRECTORIO': 'id_vivienda',
'SECUENCIA_P': 'id_hogar',
'ORDEN': 'id_persona',
'MES': 'mes',
'CLASE': 'clase',
'DPTO': 'departamento',
'FEX_C18': 'factor_expansion',
'PT': 'poblacion_total',
'P3271': 'sexo',
'P6040': 'edad',
'P6080': 'grupo_etnico',
'P3043': 'nivel_educativo',
'OCI': 'ocupados',
'P6430': 'categoria_ocupacional',
'RAMA4D_R4': 'cod_ciiu',
'INGLABO': 'ingreso_laboral'}

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.rename(columns=mapa_nombres, inplace=True)


/tmp/ipython-input-3587061779.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_empleo.rename(columns=mapa_nombres, inplace=True)


In [ ]:
# Celda de verificación (no modificar)
nombres_nuevos_esperados = list(mapa_nombres.values())
condicion_exitosa = all(col in df_empleo.columns for col in nombres_nuevos_esperados)
assert condicion_exitosa, "No todas las columnas esperadas se encuentran en el DataFrame. Revisa el diccionario o la operación de renombrado."
print("¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅")
print("¡Punto 2 completado con éxito! ✅")

¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅
¡Punto 2 completado con éxito! ✅


# Ejercicio 3: Revisión de duplicados (5%)

Un paso crucial en la limpieza es asegurar que no haya observaciones duplicadas que puedan sesgar nuestro análisis.

1. Verifica si hay filas duplicadas en `df_empleo`.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.duplicated().sum()


np.int64(0)

2. Cuenta cuántas filas duplicadas existen.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.duplicated().sum()  # Se puede usar el mismo código para las 2, ya que muestra que son 0 las observaciones duplicadas


np.int64(0)

3. Elimina las filas duplicadas, si las hay.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.drop_duplicates(inplace=True)

/tmp/ipython-input-1840243837.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_empleo.drop_duplicates(inplace=True)


4. Verifica si hay filas duplicadas específicamente en las columnas 'id_vivienda', 'id_hogar', 'id_persona'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.duplicated(subset=['id_vivienda', 'id_hogar', 'id_persona']).sum()


np.int64(0)

In [ ]:
# Celda de verificación (no modificar)
assert df_empleo.duplicated().sum() == 0, "Aún existen filas duplicadas."
print("¡Punto 3 completado con éxito! ✅")

¡Punto 3 completado con éxito! ✅


# Ejercicio 4: Valores faltantes (5%)

Los datos del mundo real raramente están completos. Aquí, trataremos los valores faltantes en la columna de ingresos.


1. Calcula el número de valores nulos (`NaN`) en la columna `ingreso_laboral` **antes** de cualquier modificación.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['ingreso_laboral'].isnull().sum()


np.int64(236576)

2. En la GEIH, un ingreso faltante a menudo significa que la persona no reportó ingresos o no aplica (ej. no trabaja). Una estrategia común es imputar estos valores con `0`. Rellena los valores `NaN` en la columna `ingreso_laboral` con `0`.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['ingreso_laboral'].fillna(0, inplace=True)


/tmp/ipython-input-3762104946.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_empleo['ingreso_laboral'].fillna(0, inplace=True)
/tmp/ipython-input-3762104946.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_empleo['ingreso_laboral'].fillna(0, inplace=True)


In [ ]:
# Celda de verificación (no modificar)
assert df_empleo['ingreso_laboral'].isnull().sum() == 0, "La columna 'ingreso_laboral' todavía tiene valores nulos."
print("¡Punto 4 completado con éxito! ✅")

¡Punto 4 completado con éxito! ✅


# Ejercicio 5: Decodificar variables (20%)



Las variables categóricas a menudo necesitan ser "traducidas" para hacerlas más entendibles para el análisis.

1. Usa los siguientes diccionarios para reemplazar los códigos numéricos por etiquetas de texto en las siguientes columnas: 'clase', 'sexo', 'grupo_etnico', 'nivel_educativo' y 'categoria_ocupacional'. Muestra los conteos de las nuevas categorías usando `.value_counts()` para verificar el cambio en cada una de las columnas.

In [ ]:
dic_clase = {1: 'Cabeceras', 2: 'Centros poblados y rural disperso'}
dic_sexo = {1: 'Hombre', 2: 'Mujer'}
dic_etnico = {1: 'Indígena', 2: 'Gitano (Rom)', 3: 'Raizal del archipiélago', 4: 'Palenquero',
              5: 'Negro, mulato (afrodescendiente)', 6: 'Ninguno de los anteriores'}
dic_nivel_educativo = {1: 'Ninguno', 2: 'Educación media', 3: 'Educación media',
                       4: 'Educación media', 5: 'Técnica profesional', 6: 'Tecnológica',
                       7: 'Universitaria', 8: 'Especialización', 9: 'Maestría',
                       10: 'Doctorado', 99: 'No sabe, no informa'}
dic_cat_ocu = {1: 'Asalariado', 2: 'Asalariado', 3: 'Asalariado',
               8: 'Asalariado', 4: 'Independiente', 5: 'Independiente',
               9: 'Independiente', 6: 'Trabajador sin remuneración',
               7: 'Trabajador sin remuneración'}

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['clase'] = df_empleo['clase'].replace(dic_clase)
df_empleo['sexo'] = df_empleo['sexo'].replace(dic_sexo)
df_empleo['grupo_etnico'] = df_empleo['grupo_etnico'].replace(dic_etnico)
df_empleo['nivel_educativo'] = df_empleo['nivel_educativo'].replace(dic_nivel_educativo)
df_empleo['categoria_ocupacional'] = df_empleo['categoria_ocupacional'].replace(dic_cat_ocu)

print("Value counts after mapping:")
print("\nClase:")
print(df_empleo['clase'].value_counts(dropna=False))
print("\nSexo:")
print(df_empleo['sexo'].value_counts(dropna=False))
print("\nGrupo Etnico:")
print(df_empleo['grupo_etnico'].value_counts(dropna=False))
print("\nNivel Educativo:")
print(df_empleo['nivel_educativo'].value_counts(dropna=False))
print("\nCategoria Ocupacional:")
print(df_empleo['categoria_ocupacional'].value_counts(dropna=False))

/tmp/ipython-input-4083600672.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_empleo['clase'] = df_empleo['clase'].replace(dic_clase)
/tmp/ipython-input-4083600672.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_empleo['sexo'] = df_empleo['sexo'].replace(dic_sexo)
/tmp/ipython-input-4083600672.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pa

Value counts after mapping:

Clase:
clase
Cabeceras                            356225
Centros poblados y rural disperso     52198
Name: count, dtype: int64

Sexo:
sexo
Mujer     218100
Hombre    190323
Name: count, dtype: int64

Grupo Etnico:
grupo_etnico
Ninguno de los anteriores           351537
Negro, mulato (afrodescendiente)     38516
Indígena                             17191
Raizal del archipiélago               1004
Palenquero                             126
Gitano (Rom)                            49
Name: count, dtype: int64

Nivel Educativo:
nivel_educativo
NaN                    199969
Educación media        117166
Universitaria           32763
Técnica profesional     24196
Ninguno                 12797
Tecnológica             10561
Especialización          7584
Maestría                 3067
Doctorado                 317
No sabe, no informa         3
Name: count, dtype: int64

Categoria Ocupacional:
categoria_ocupacional
NaN                            232247
Asalariado      

/tmp/ipython-input-4083600672.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_empleo['categoria_ocupacional'] = df_empleo['categoria_ocupacional'].replace(dic_cat_ocu)


In [ ]:
# Celda de verificación (no modificar)
mapeos_a_verificar = {
    'clase': dic_clase,
    'sexo': dic_sexo,
    'grupo_etnico': dic_etnico,
    'nivel_educativo': dic_nivel_educativo,
    'categoria_ocupacional': dic_cat_ocu
}
for columna, diccionario in mapeos_a_verificar.items():
    es_numerica = pd.api.types.is_numeric_dtype(df_empleo[columna])
    assert not es_numerica, f"La columna '{columna}' todavía tiene un tipo de dato numérico. El reemplazo falló."
    valor_texto_ejemplo = list(diccionario.values())[0]
    categorias_unicas = df_empleo[columna].unique()
    assert valor_texto_ejemplo in categorias_unicas, f"No se encontró el valor '{valor_texto_ejemplo}' en la columna '{columna}'."
    valor_numerico_ejemplo = list(diccionario.keys())[0]
    assert valor_numerico_ejemplo not in categorias_unicas, f"El código numérico '{valor_numerico_ejemplo}' todavía existe en la columna '{columna}'."
print("¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.")

¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.


2. Utiliza el archivo "2025-01-14 Correlativa CIIU Rev4.xlsx", hoja 'CIIU 2022', para agregar al df_empleo las descripciones de los códigos CIIU a 4 dígitos. En df_empleo la columna con los códigos CIIU se llama 'cod_ciiu'. Si se agregan nuevas columnas al dataframe, distintas a la descripción del CIIU, debes borrarlas.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_ciiu = pd.read_excel('/content/2025-01-14 Correlativa CIIU Rev4.xlsx', sheet_name='CIIU 2022')
df_ciiu['RAMA4D_R4'] = df_ciiu['RAMA4D_R4'].astype(str)
df_empleo = pd.merge(df_empleo, df_ciiu[['RAMA4D_R4', 'DESCRIPCION_CIIU']], left_on='cod_ciiu', right_on='RAMA4D_R4', how='left')
df_empleo = df_empleo.drop(columns=['RAMA4D_R4'])
df_empleo.head()

,id_vivienda,id_hogar,id_persona,mes,clase,departamento,factor_expansion,poblacion_total,sexo,edad,grupo_etnico,nivel_educativo,ocupados,categoria_ocupacional,cod_ciiu,ingreso_laboral,DESCRIPCION_CIIU
0,8207971,1,1,5,Centros poblados y rural disperso,08,75.795526,1,Hombre,33,Ninguno de los anteriores,Educación media,1.0,Asalariado,9609,3400000.0,Otras actividades de servicios personales n.c.p.
1,8207971,1,2,5,Centros poblados y rural disperso,08,75.795526,1,Hombre,32,Ninguno de los anteriores,Técnica profesional,1.0,Asalariado,9609,1400000.0,Otras actividades de servicios personales n.c.p.
2,8207972,1,1,5,Centros poblados y rural disperso,08,65.131042,1,Hombre,26,Ninguno de los anteriores,Técnica profesional,1.0,Independiente,4921,750000.0,Transporte de pasajeros
3,8207972,1,2,5,Centros poblados y rural disperso,08,65.131042,1,Mujer,24,Ninguno de los anteriores,Universitaria,1.0,Asalariado,4773,840000.0,Comercio al por menor de productos farmacéutic...
4,8207973,1,1,5,Centros poblados y rural disperso,08,33.177818,1,Hombre,75,Ninguno de los anteriores,NaN,1.0,Independiente,4111,600000.0,Construcción de edificios residenciales


In [ ]:
# Celda de verificación (no modificar)
assert 'DESCRIPCION_CIIU' in df_empleo.columns, \
    f"La columna '{'DESCRIPCION_CIIU'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['DESCRIPCION_CIIU'].notna().sum() > 0, \
    f"La columna '{'DESCRIPCION_CIIU'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")

✅ Verificación del merge exitosa.


3. Utiliza el archivo "2025-01-14 Divipola.xlsx", hoja 'Departamentos', para agregar al df_empleo los nombres de los departamentos. En df_empleo la columna con los códigos Divipola de los departamentos se llama 'departamento'. Si se agregan nuevas columnas al dataframe, distintas al nombre del departamento, debes borrarlas.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_divipola = pd.read_excel('/content/2025-01-14 DIVIPOLA.xlsx', sheet_name='Departamentos')
df_divipola = df_divipola.iloc[8:].copy()
df_divipola = df_divipola.rename(columns={'Unnamed: 0': 'departamento', 'Unnamed: 1': 'Nombre'})
df_divipola['departamento'] = df_divipola['departamento'].astype(str)
df_divipola = df_divipola[['departamento', 'Nombre']]
df_empleo = pd.merge(df_empleo, df_divipola, on='departamento', how='left')
df_empleo.head()

,id_vivienda,id_hogar,id_persona,mes,clase,departamento,factor_expansion,poblacion_total,sexo,edad,grupo_etnico,nivel_educativo,ocupados,categoria_ocupacional,cod_ciiu,ingreso_laboral,DESCRIPCION_CIIU,Nombre
0,8207971,1,1,5,Centros poblados y rural disperso,08,75.795526,1,Hombre,33,Ninguno de los anteriores,Educación media,1.0,Asalariado,9609,3400000.0,Otras actividades de servicios personales n.c.p.,ATLÁNTICO
1,8207971,1,2,5,Centros poblados y rural disperso,08,75.795526,1,Hombre,32,Ninguno de los anteriores,Técnica profesional,1.0,Asalariado,9609,1400000.0,Otras actividades de servicios personales n.c.p.,ATLÁNTICO
2,8207972,1,1,5,Centros poblados y rural disperso,08,65.131042,1,Hombre,26,Ninguno de los anteriores,Técnica profesional,1.0,Independiente,4921,750000.0,Transporte de pasajeros,ATLÁNTICO
3,8207972,1,2,5,Centros poblados y rural disperso,08,65.131042,1,Mujer,24,Ninguno de los anteriores,Universitaria,1.0,Asalariado,4773,840000.0,Comercio al por menor de productos farmacéutic...,ATLÁNTICO
4,8207973,1,1,5,Centros poblados y rural disperso,08,33.177818,1,Hombre,75,Ninguno de los anteriores,NaN,1.0,Independiente,4111,600000.0,Construcción de edificios residenciales,ATLÁNTICO


In [ ]:
# Celda de verificación (no modificar)
assert 'Nombre' in df_empleo.columns, \
    f"La columna '{'Nombre'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['Nombre'].notna().sum() > 0, \
    f"La columna '{'Nombre'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 18, f"Error en las columnas. Se esperaban 18, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 5 completado con éxito! ✅")

✅ Verificación del merge exitosa.
¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 5 completado con éxito! ✅


# Ejercicio 6: Crear nuevas variables y reorganizar columnas (10%)

La GEIH cuenta con un factor de expansión que permite extrapolar los datos de los encuestados a totales para la población colombiana. Usar este factor de expansión, permite calcular la población total del país, la población en edad de trabajar, el número de ocupados, el número de desempleados, entre otros datos.

1. Crea una nueva columna llamada "fex", la cual corresponderá a la columna 'factor_expansion' dividida en 6 (se divide en 6, ya que estamos tomando el dato de 6 meses).

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['fex']=df_empleo['factor_expansion']/6


2. Crea una nueva columna llamada 'poblacion_total_ex', la cual corresponderá a la columna 'poblacion_total' multiplicada por la columna 'fex'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['poblacion_total_ex']=df_empleo['poblacion_total']*df_empleo['fex']


3. Crea una nueva columna llamada 'ocupados_ex', la cual corresponderá a la columna 'ocupados' multiplicada por la columna 'fex'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['ocupados_ex']=df_empleo['ocupados']*df_empleo['fex']


4. Crea una nueva columna llamada 'joven', la cual debe decir 'Población joven' si la persona encuestada tiene entre 15 y 28 años (15 y 28 incluidos). De lo contrario, debe decir 'Resto de población'.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo['joven'] = df_empleo['edad'].apply(lambda x: 'Población joven' if 15 <= x <= 28 else 'Resto de población')


5. Reorganiza las columnas con el siguiente orden: 'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'Nombre', 'factor_expansion', 'fex', 'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'

In [ ]:
### TU CÓDIGO AQUÍ ###
import pandas as pd
if isinstance(df_empleo,list):
     df_empleo = pd.DataFrame(df_empleo)


In [ ]:
### TU CÓDIGO AQUÍ ###
orden_deseado = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU',
    'ingreso_laboral'
]

df_empleo = df_empleo[orden_deseado]
print(df_empleo.shape)
df_empleo.head()

(408423, 22)


,id_vivienda,id_hogar,id_persona,mes,clase,departamento,Nombre,factor_expansion,fex,poblacion_total,...,edad,joven,grupo_etnico,nivel_educativo,ocupados,ocupados_ex,categoria_ocupacional,cod_ciiu,DESCRIPCION_CIIU,ingreso_laboral
0,8207971,1,1,5,Centros poblados y rural disperso,08,ATLÁNTICO,75.795526,12.632588,1,...,33,Resto de población,Ninguno de los anteriores,Educación media,1.0,12.632588,Asalariado,9609,Otras actividades de servicios personales n.c.p.,3400000.0
1,8207971,1,2,5,Centros poblados y rural disperso,08,ATLÁNTICO,75.795526,12.632588,1,...,32,Resto de población,Ninguno de los anteriores,Técnica profesional,1.0,12.632588,Asalariado,9609,Otras actividades de servicios personales n.c.p.,1400000.0
2,8207972,1,1,5,Centros poblados y rural disperso,08,ATLÁNTICO,65.131042,10.855174,1,...,26,Población joven,Ninguno de los anteriores,Técnica profesional,1.0,10.855174,Independiente,4921,Transporte de pasajeros,750000.0
3,8207972,1,2,5,Centros poblados y rural disperso,08,ATLÁNTICO,65.131042,10.855174,1,...,24,Población joven,Ninguno de los anteriores,Universitaria,1.0,10.855174,Asalariado,4773,Comercio al por menor de productos farmacéutic...,840000.0
4,8207973,1,1,5,Centros poblados y rural disperso,08,ATLÁNTICO,33.177818,5.529636,1,...,75,Resto de población,Ninguno de los anteriores,NaN,1.0,5.529636,Independiente,4111,Construcción de edificios residenciales,600000.0


In [ ]:
# Celda de verificación (no modificar)
orden_esperado = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'
]
orden_actual = df_empleo.columns.tolist()
assert orden_actual == orden_esperado, \
    f"""El orden de las columnas es incorrecto.
    - Orden Esperado: {orden_esperado}
    - Orden Encontrado: {orden_actual}"""
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 22, f"Error en las columnas. Se esperaban 22, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 6 completado con éxito! ✅")

¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 6 completado con éxito! ✅


# Ejercicio 7: Cambiar tipo de dato (5%)



Para realizar cálculos y análisis correctamente, las columnas deben tener el tipo de dato adecuado (numérico, categórico, etc.).

1.  Muestra en pantalla los tipos de datos actuales de `df_empleo`.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo.dtypes


,0
id_vivienda,int64
id_hogar,int64
id_persona,int64
mes,int64
clase,object
departamento,object
Nombre,object
factor_expansion,float64
fex,float64
poblacion_total,int64


2.  Las siguientes columnas representan categorías: 'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU'. Conviértelas al tipo `category` para optimizar el uso de memoria y facilitar análisis futuros.

In [ ]:
### TU CÓDIGO AQUÍ ###
columnas_categorizadas= [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]

for col in columnas_categorizadas:
    df_empleo[col] = df_empleo[col].astype('category')

In [ ]:
# Celda de verificación (no modificar)
columnas_categoricas_esperadas = [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]
for columna in columnas_categoricas_esperadas:
    assert columna in df_empleo.columns, \
        f"La columna '{columna}' no fue encontrada en el DataFrame."
    es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])
    assert es_categoria, \
        f"La columna '{columna}' no fue convertida al tipo 'category'. Su tipo actual es {df_empleo[columna].dtype}."
print("¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.")
print("\n¡Punto 7 completado con éxito! ✅")

¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.

¡Punto 7 completado con éxito! ✅


/tmp/ipython-input-416816491.py:10: DeprecationWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, pd.CategoricalDtype) instead
  es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])


# Ejercicio 8: Ordenar el dataframe (5%)

1. Ordena el dataframe por la columna "edad" de mayor a menor. Observa la columna 'edad' y escribe en una celda de markdown cuál fue la edad mínima y la edad máxima de las personas que respondieron la encuesta de la GEIH.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_empleo.sort_values(by='edad', ascending=False)
df_empleo.head()


,id_vivienda,id_hogar,id_persona,mes,clase,departamento,Nombre,factor_expansion,fex,poblacion_total,...,edad,joven,grupo_etnico,nivel_educativo,ocupados,ocupados_ex,categoria_ocupacional,cod_ciiu,DESCRIPCION_CIIU,ingreso_laboral
209236,8146261,1,3,3,Cabeceras,19,CAUCA,134.149141,22.358190,1,...,106,Resto de población,Ninguno de los anteriores,NaN,NaN,NaN,NaN,NaN,NaN,0.0
255709,8165410,1,3,3,Cabeceras,27,CHOCÓ,482.953528,80.492255,1,...,106,Resto de población,"Negro, mulato (afrodescendiente)",NaN,NaN,NaN,NaN,NaN,NaN,0.0
136404,8173637,1,2,4,Centros poblados y rural disperso,08,ATLÁNTICO,190.477567,31.746261,1,...,106,Resto de población,Ninguno de los anteriores,NaN,NaN,NaN,NaN,NaN,NaN,0.0
100871,8129316,1,4,2,Cabeceras,70,SUCRE,409.240792,68.206799,1,...,105,Resto de población,Indígena,NaN,NaN,NaN,NaN,NaN,NaN,0.0
163244,8190663,1,1,4,Cabeceras,73,TOLIMA,219.393600,36.565600,1,...,105,Resto de población,Ninguno de los anteriores,NaN,NaN,NaN,NaN,NaN,NaN,0.0


In [ ]:
df_empleo['edad'].max()

106

In [ ]:
df_empleo['edad'].min() # Hay valores == 0, es por eso que muestra edad mínima 0

0

In [ ]:
(df_empleo['edad'] == 0).sum()

np.int64(3441)

Escribe tu análisis aquí:
Revisando la columna edad se observó que existían registros con valor 0,lo cual no deberí ser posible. Sin embargo hay 3441 casos con edad == 0. Se podría limpiar la base para eliminar esos valores de 0 y con ello determinar cual sería la edad mínima distinta de 0 en la encuesta.

In [ ]:
# Celda de verificación (no modificar)
es_descendiente = df_empleo['edad'].is_monotonic_decreasing
assert es_descendiente, "La columna 'edad' no está ordenada correctamente de mayor a menor."
print("\n¡Punto 8 completado con éxito! ✅")


¡Punto 8 completado con éxito! ✅


# Ejercicio 9: Groupby (20%)

Agrupar los datos por categorías, nos permite realizar análisis más profundos y obtener insights valiosos, que a simple vista no vemos en la base de datos.

1. Crea un nuevo dataframe llamado "df_sexo" donde se agrupe por 'sexo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de sexo en una columna.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_sexo = df_empleo.groupby('sexo').agg({'poblacion_total_ex': 'sum', 'ocupados_ex': 'sum'}).reset_index()
df_sexo['porcentaje_ocupados'] = df_sexo['ocupados_ex'] / df_sexo['poblacion_total_ex']

/tmp/ipython-input-3305567648.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_sexo = df_empleo.groupby('sexo').agg({'poblacion_total_ex': 'sum', 'ocupados_ex': 'sum'}).reset_index()


In [ ]:
df_sexo

,sexo,poblacion_total_ex,ocupados_ex,porcentaje_ocupados
0,Hombre,2.525511e+07,1.379736e+07,0.546320
1,Mujer,2.668224e+07,9.707732e+06,0.363827


In [ ]:
# Celda de verificación (no modificar)
import numpy as np
dimensiones_esperadas = (2, 4)
assert df_sexo.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas}, pero se encontró {df_sexo.shape}. ¿Aplicaste .reset_index()?"
assert 'sexo' in df_sexo.columns, \
    "La columna 'sexo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
assert not isinstance(df_sexo.index, pd.Index) or df_sexo.index.name != 'sexo', \
    "El índice del DataFrame sigue siendo 'sexo'. Asegúrate de haber aplicado .reset_index()."
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


2. Crea un nuevo dataframe llamado "df_educacion" donde se agrupe por 'nivel_educativo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de nivel educativo en una columna.

In [ ]:
### TU CÓDIGO AQUÍ ###
df_educacion = df_empleo.groupby('nivel_educativo').agg({'poblacion_total_ex': 'sum', 'ocupados_ex': 'sum'}).reset_index()
df_educacion['porcentaje_ocupados'] = df_educacion['ocupados_ex'] / df_educacion['poblacion_total_ex']

/tmp/ipython-input-1994326004.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_educacion = df_empleo.groupby('nivel_educativo').agg({'poblacion_total_ex': 'sum', 'ocupados_ex': 'sum'}).reset_index()


In [ ]:
df_educacion

,nivel_educativo,poblacion_total_ex,ocupados_ex,porcentaje_ocupados
0,Doctorado,4.664918e+04,3.995047e+04,0.856402
1,Educación media,1.432753e+07,8.605758e+06,0.600645
2,Especialización,9.652363e+05,7.777595e+05,0.805771
3,Maestría,4.486660e+05,3.869646e+05,0.862478
4,Ninguno,1.710908e+06,3.487339e+05,0.203830
5,"No sabe, no informa",9.477569e+01,3.760248e+01,0.396752
6,Tecnológica,1.392853e+06,1.040970e+06,0.747365
7,Técnica profesional,2.781981e+06,1.953369e+06,0.702150
8,Universitaria,3.731910e+06,2.759725e+06,0.739494


In [ ]:
# --- Celda de Verificación (No modificar) ---
import numpy as np
n_niveles_educativos = df_empleo['nivel_educativo'].nunique()
dimensiones_esperadas = (n_niveles_educativos, 4)
assert df_educacion.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas} (filas, columnas), pero se encontró {df_educacion.shape}."
columnas_esperadas = {'nivel_educativo', 'poblacion_total_ex', 'ocupados_ex', 'porcentaje_ocupados'}
assert set(df_educacion.columns) == columnas_esperadas, \
    f"Los nombres de las columnas no son correctos. Se esperaban {columnas_esperadas}."
assert 'nivel_educativo' in df_educacion.columns, \
    "La columna 'nivel_educativo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


3. En una celda de Markdown escribe qué conclusiones puedes sacar de la información obtenida en df_sexo y df_educacion.

Escribe tu análisis aquí: De acuerdo a la info obtenida de df_sexo podemos decir que los hombres tienen una mayor tasa de ocupación que las mujeres, lo que confirma una brecha de género en temas de acceso al empleo, a pesar de que las mujeres tienen una población mayor que la de los hombres.

Respecto a la info obtenida en df_eduacion podemos ver si comparamos ligeramente los datos, que el nivel educativo y la tasa de ocupación tienen una relación que tiende a ser positiva. Los niveles de Maestría y Doctorado son los que mas porcentaje de ocupados tienen con 86% y 85% respectivamente. Mientras que los que no tienen educacion representan un 20%.

In [ ]:
# Celda de verificación (no modificar)
print("\n¡Punto 9 completado con éxito! ✅")


¡Punto 9 completado con éxito! ✅


# Ejercicio 10: Exportar datos (5%)

Muchas veces requerimos exportar los resultados de nuestros análisis a Excel para compartir la información con otras personas.

Exporta los dataframes df_sexo y df_educacion a un Excel llamado 'Analisis_ocupados.xlsx'. Cada dataframe debe quedar en una hoja diferente. La hoja de 'df_sexo' se debe llamar 'sexo' y la hoja de 'df_educacion' se debe llamar 'educacion'.

In [ ]:
### TU CÓDIGO AQUÍ ###
with pd.ExcelWriter('Analisis_ocupados.xlsx') as writer:
    df_sexo.to_excel(writer, sheet_name='sexo', index=False)


In [ ]:
# Celda de verificación (no modificar)
import os
import pandas as pd
nombre_archivo = 'Analisis_ocupados.xlsx'
assert os.path.exists(nombre_archivo), \
    f"No se encontró el archivo '{nombre_archivo}'. ¿Lo exportaste con el nombre correcto?"
try:
    xls = pd.ExcelFile(nombre_archivo)
    nombres_hojas_encontrados = xls.sheet_names
    nombres_hojas_esperados = ['sexo', 'educacion']
    assert set(nombres_hojas_encontrados) == set(nombres_hojas_esperados), \
        f"Los nombres de las hojas no son correctos. Se esperaban {nombres_hojas_esperados} pero se encontraron {nombres_hojas_encontrados}."
    print("¡Verificación completada con éxito! ✅")
except Exception as e:
    print(f"Ocurrió un error al verificar el archivo: {e}")
print("\n¡Punto 10 completado con éxito! ✅")

Ocurrió un error al verificar el archivo: Los nombres de las hojas no son correctos. Se esperaban ['sexo', 'educacion'] pero se encontraron ['sexo'].

¡Punto 10 completado con éxito! ✅
